src/test/net_tests.cpp

```diff
@@ -704,4 +705,82 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
     BOOST_CHECK(!IsLocal(addr));
 }
 
+void message_serialize_deserialize_test(bool v2, const std::vector<CSerializedNetMsg>& test_msgs)
+{
+    // use 32 byte keys with all zeros
+    CPrivKey k1(32, 0);
+    CPrivKey k2(32, 0);
+
+    // construct the serializers
+    std::unique_ptr<TransportSerializer> serializer;
+    std::unique_ptr<TransportDeserializer> deserializer;
+
+    if (v2) {
+        serializer = std::make_unique<V2TransportSerializer>(V2TransportSerializer(k1, k2));
+        deserializer = std::make_unique<V2TransportDeserializer>(V2TransportDeserializer((NodeId)0, k1, k2));
+    } else {
+        serializer = std::make_unique<V1TransportSerializer>(V1TransportSerializer());
+        deserializer = std::make_unique<V1TransportDeserializer>(V1TransportDeserializer(Params(), (NodeId)0, SER_NETWORK, INIT_PROTO_VERSION));
+    }
+    // run a couple of times through all messages with the same AEAD instance
+    for (unsigned int i = 0; i < 100; i++) {
+        for (const CSerializedNetMsg& msg_orig : test_msgs) {
+            // bypass the copy protection
+            CSerializedNetMsg msg;
+            msg.data = msg_orig.data;
+            msg.m_type = msg_orig.m_type;
+            size_t raw_msg_size = msg.data.size();
+
+            std::vector<unsigned char> serialized_header;
+            serializer->prepareForTransport(msg, serialized_header);
+
+            // read two times
+            //  first: read header
+            size_t read_bytes = 0;
+            Span<const uint8_t> span_header(serialized_header.data(), serialized_header.size());
+            if (serialized_header.size() > 0) read_bytes += deserializer->Read(span_header);
+            //  second: read the encrypted payload (if required)
+            Span<const uint8_t> span_msg(msg.data.data(), msg.data.size());
+            if (msg.data.size() > 0) read_bytes += deserializer->Read(span_msg);
+            if (msg.data.size() > read_bytes && msg.data.size() - read_bytes > 0) {
```

Comment:
```suggestion
            if (msg.data.size() > read_bytes) {
```

Reply:
Oh, that was quite silly. Thanks for reviewing closely. Done.